normandy/recipes/utils.py

```diff
@@ -0,0 +1,48 @@
+import hashlib
+
+
+def fraction_to_key(frac):
+    """Map from the range [0, 1] to [0, max(sha256)]. The result is a string."""
+    # SHA 256 hashes are 64 hexadecimal numbers. The largest possible SHA 256
+    # hash is 2^64 - 1
+
+    if frac < 0 or frac > 1:
+        raise ValueError('frac must be between 0 and 1 inclusive (got {})'.format(frac))
+
+    mult = 2 ** 256 - 1
+    in_decimal = int(frac * mult)
+
+    assert in_decimal >= 0
+
+    hex_digits = hex(in_decimal)[2:]  # Strip off leading "0x"
+    padded = "{:0<64}".format(hex_digits)
+
+    # Saturate at 2**256 - 1
+    if len(padded) > 64:
+        return 'f' * 64
+    else:
+        return padded
+
+
+def deterministic_sample(rate, inputs):
+    """
+    Deterministically choose True or False based for a set of inputs.
+
+    Internally, converts `rate` into a point in the sha256 hash space. If the
+    hash of `inputs` is less than that point, it returns True.
+
+    @parameter rate The probability of returning True
```

Comment:
Oops. This is a bastardization of Javadoc that I was using in JS. I'll switch this over to [http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists](Sphinx style).